torch/distributed/checkpoint/utils.py

```diff
@@ -46,6 +47,13 @@ def _get_failure_dict(
         {i: err for i, err in enumerate(results) if _is_wrapped_exception(err)},
     )
 
+def _all_gather_keys(local_dict: Dict[Any, Any]) -> List[Any]:
+    """Gathers all keys, and returns them sorted."""
+    keys = list(local_dict.keys())
+    gathered_keys: List[List[Any]] = [None] * dist.get_world_size()  # type: ignore[list-item]
+
+    dist.all_gather_object(gathered_keys, keys)
+    return sorted(set(itertools.chain.from_iterable(gathered_keys)))
```

Comment:
Can we assert if `set(local_dict)` does not match `set(sorted(...))`? We should only solve the oder problem here but if there are missing keys, we should assert.

Reply:
Any strong opinions on warning instead of an assert? @fegin @wz337

Reply:
I went with a warning